frontend/web/src/assets/lib/jsc3d/jsc3d.webgl.js

```diff
@@ -0,0 +1,1202 @@
+					normals[j + 4] = nbuf[n1 + 1];
+					normals[j + 5] = nbuf[n1 + 2];
+					normals[j + 6] = nbuf[n2    ];
+					normals[j + 7] = nbuf[n2 + 1];
+					normals[j + 8] = nbuf[n2 + 2];
+				}
+
+				if(hasTextureCoords) {
+					t0 = tibuf[i    ] * 2;
+					t1 = tibuf[i + 1] * 2;
+					t2 = tibuf[i + 2] * 2;
+					texcoords[k    ] = tbuf[t0    ];
+					texcoords[k + 1] = tbuf[t0 + 1];
+					texcoords[k + 2] = tbuf[t1    ];
+					texcoords[k + 3] = tbuf[t1 + 1];
+					texcoords[k + 4] = tbuf[t2    ];
+					texcoords[k + 5] = tbuf[t2 + 1];
+				}
+			}
+		}
+		else {
+			coords = [];
+			normals = [];
+			if(hasTextureCoords)
+				texcoords = [];
+			var v0, v1, v2;
+			var n0, n1, n2;
+			var t0, t1, t2;
+			for(var i=0, j=0; i<numOfFaces; i++) {
+				v0 = ibuf[j    ] * 3;
+				v1 = ibuf[j + 1] * 3;
+				n0 = nibuf[j    ] * 3;
+				n1 = nibuf[j + 1] * 3;
+				if(hasTextureCoords) {
+					t0 = tibuf[j    ] * 2;
+					t1 = tibuf[j + 1] * 2;
+				}
+				j += 2;
+				while(ibuf[j] >= 0) {
+					v2 = ibuf[j] * 3;
+					coords.push(vbuf[v0], vbuf[v0 + 1], vbuf[v0 + 2], vbuf[v1], vbuf[v1 + 1], vbuf[v1 + 2], vbuf[v2], vbuf[v2 + 1], vbuf[v2 + 2]);
+					v1 = v2;
+					if(needFlat) {
+						n0 = i * 3;
+						normals.push(fnbuf[n0], fnbuf[n0 + 1], fnbuf[n0 + 2], fnbuf[n0], fnbuf[n0 + 1], fnbuf[n0 + 2], fnbuf[n0], fnbuf[n0 + 1], fnbuf[n0 + 2]);
+					}
+					else {
+						n2 = nibuf[j] * 3;
+						normals.push(nbuf[n0], nbuf[n0 + 1], nbuf[n0 + 2], nbuf[n1], nbuf[n1 + 1], nbuf[n1 + 2], nbuf[n2], nbuf[n2 + 1], nbuf[n2 + 2]);
+						n1 = n2;
+					}
+					if(hasTextureCoords) {
+						t2 = tibuf[j] * 2;
+						texcoords.push(tbuf[t0], tbuf[t0 + 1], tbuf[t1], tbuf[t1 + 1], tbuf[t2], tbuf[t2 + 1]);
+						t1 = t2;
+					}
+					j++;
+				}
+				j++;
+			}
+			coords = new Float32Array(coords);
+			normals = new Float32Array(normals);
+			if(hasTextureCoords)
+				texcoords = new Float32Array(texcoords);
+		}
+
+		mesh.compiled.coords = gl.createBuffer();
+		gl.bindBuffer(gl.ARRAY_BUFFER, mesh.compiled.coords);
+		gl.bufferData(gl.ARRAY_BUFFER, coords, gl.STATIC_DRAW);
+		mesh.compiled.normals = gl.createBuffer();
+		gl.bindBuffer(gl.ARRAY_BUFFER, mesh.compiled.normals);
+		gl.bufferData(gl.ARRAY_BUFFER, normals, gl.STATIC_DRAW);
+		if(hasTextureCoords) {
+			mesh.compiled.texcoords = gl.createBuffer();
+			gl.bindBuffer(gl.ARRAY_BUFFER, mesh.compiled.texcoords);
+			gl.bufferData(gl.ARRAY_BUFFER, texcoords, gl.STATIC_DRAW);
+		}
+		gl.bindBuffer(gl.ARRAY_BUFFER, null);
+
+		mesh.compiled.faceCount = numOfFaces;
+		mesh.compiled.coordCount = coords.length / 3;
+	}
+	else {
+		/*
+		 * Do not need to rebuild, just update normal data.
+		 */
+
+		var isFlat = (mesh.compiled.remderMode == 'flat') || (mesh.compiled.remderMode == 'textureflat');
+		if(isFlat != needFlat) {
+			var normals;
+			if(hasTrianglesOnly) {
+				normals = new Float32Array(9 * numOfFaces);
+				var n0, n1, n2
+				for(var i=0, j=0, faceIndex=0; i<ibuf.length; i+=4, j+=9, faceIndex++) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "faceIndex" as this name is already used in declaration at line 927. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "i" as this name is already used in declaration at line 927. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "j" as this name is already used in declaration at line 927. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)